crates/doc/src/reduce/strategy.rs

```diff
@@ -301,6 +309,54 @@ impl Strategy {
         }
     }
 
+    fn json_schema_merge<'alloc, L: AsNode, R: AsNode>(
+        cur: Cursor<'alloc, '_, '_, '_, '_, L, R>,
+    ) -> Result<HeapNode<'alloc>> {
+        let Cursor {
+            tape,
+            loc,
+            full: _,
+            lhs,
+            rhs,
+            alloc,
+        } = cur;
+
+        let (lhs, rhs) = (lhs.into_heap_node(alloc), rhs.into_heap_node(alloc));
+
+        // Precompute this up here because we're going to consume `rhs` later on
+        let rhs_tape_len = count_nodes_heap(&rhs);
+
+        // Ensure that we're working with objects on both sides
+        // Question: Should we actually relax this to support
+        // reducing valid schemas like "true" and "false"?
+        let (
+            lhs @ HeapNode::Object(_),
+            rhs @ HeapNode::Object(_)
+        ) = (lhs, rhs) else {
+            return Err(Error::with_location(Error::JsonSchemaMergeWrongType { detail: None }, loc) )
+        };
+
+        let left = shape_from_node(lhs).map_err(|e| Error::with_location(e, loc))?;
+        let right = shape_from_node(rhs).map_err(|e| Error::with_location(e, loc))?;
+
+        // Union together the LHS and RHS, and convert back from `Shape` into `HeapNode`.
+        let merged_doc =
+            serde_json::to_value(&SchemaBuilder::new(Shape::union(left, right)).root_schema())
```

Comment:
do you also need to call the routine to trim excessive numbers of properties?

Reply:
Wow, yes.